### PR TITLE
feat(auto-backup): auto backup in public cloud

### PR DESCRIPTION
**Pull request by radireddy**
support to create and schedule auto backup for public cloud servers

added below new features
1. Workflow on boarding page
2. Workflow list and executions page
3. Create new workflow page
4. Delete workflow page

MBP-456

### CODE REVIEW

File: packages/manager/modules/pci/src/projects/project/instances/instances.controller.js [hunk 1]
```diff
@@ -6,13 +6,15 @@ export default class CloudProjectComputeInfrastructureListCtrl {
...
     $q,
+    $state,
     $translate,
     CucCloudMessage,
     CucRegionService,
     ovhUserPref,
     PciProjectsProjectInstanceService,
   ) {
     this.$q = $q;
+    this.$state = $state;
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/instances/instances.controller.js]: Please avois injecting `$state` in controller and prefer go through `resolve` and component `bindings`

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.html [hunk 2]
```diff
@@ -0,0 +1,57 @@
+<h1 data-translate="pci_workflow_add"></h1>
+<cui-message-container data-messages="$ctrl.messages"></cui-message-container>
+<oui-stepper>
+    <oui-step-form
+        name="workflow_type"
+        data-header="{{:: 'pci_workflow_create_type_title' | translate }}"
+        data-editable="!$ctrl.isAdding"
+        data-valid="$ctrl.workflow.type"
+        data-navigation="$ctrl.workflow.type">
+        <workflow-type
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/add.html]: Should  be prefixed something like `pci-project-instance-workflow-type`

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  schedule,
+  generalInfo,
+  'ngOvhCloudUniverseComponents',
+  'ngOvhSwimmingPoll',
+  'oui',
+  'ovh-api-services',
+])
+  .config(routing)
+  .component('ovhManagerPciProjectWorkflowAdd', component)
+  .service('PciProjectRegions', PciProjectRegions)
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: ⚠️ This will cause conflicts with https://github.com/ovh-ux/manager/pull/961

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.routing.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      component: 'ovhManagerPciProjectWorkflowAdd',
+      params: {
+        selectedInstance: null,
+      },
+      redirectTo: transition => transition
+        .injector()
+        .getAsync('instances')
+        .then(instances => (instances.length === 0 ? { state: 'pci.projects.project.instances' } : false)),
+      resolve: {
+        selectedInstance: /* @ngInject */ $stateParams => $stateParams.selectedInstance,
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/add.routing.js]: Prefer `$transitions.param()`

File: packages/manager/modules/pci/src/projects/project/workflow/add/general/general.html [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    <p> <span data-translate="pci_workflow_create_price_title"></span> 
+        <strong data-ng-if="$ctrl.price">
+            <span data-translate="pci_workflow_create_price"
+                data-translate-values="{ price: $ctrl.price.totalPrice.text }"></span>
+        </strong>
+        <strong data-ng-if="!$ctrl.price" class="text-danger">
+            <span data-translate="pci_workflow_create_price_not_available"></span>
+        </strong>
+    </p>
+</form>
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/general/general.html]: Form with no `submit` action ?

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    this.weekdaysAlias = {
+      sun: '0',
+      mon: '1',
+      tue: '2',
+      wed: '3',
+      thu: '4',
+      fri: '5',
+      sat: '6',
+    };
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: Maybe those could be handled by `moment` ?

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
+import map from 'lodash/map';
+
+export default class CronValidator {
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: There's already a validator existit in `web-universe-components` I think.
this logic is already complicated, we should probably try to not duplicate it

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.controller.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+export default class {
+  $onInit() {
+    this.minutesInvalid = false;
+    this.hoursInvalid = false;
+    this.domInvalid = false;
+    this.monthInvalid = false;
+    this.dowInvalid = false;
+    this.cronValidator = new CronValidator();
+    this.WORKFLOW_SCHEDULES = [{
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.controller.js]: This should be in a `constant` file maybe ?

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+            data-description="{{ ::'pci_workflow_create_schedule_' + schedule.name + '_description' | translate }}"
+            data-model="$ctrl.schedule"
+            data-label="{{ ::'pci_workflow_create_schedule_' + schedule.name + '_title' | translate }}"
+            data-values="[schedule]"
+            data-variant="light"
+            data-required>
+        </oui-select-picker>
+    </div>
+    <div class="row" data-ng-if="$ctrl.schedule && $ctrl.schedule.name === 'custom'">
+        <div class="col-md-12">
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html]: No need for this then 
```suggestion
        <div>
```

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+            <h3 data-translate="pci_workflow_create_cron_title" class="d-inline-block pr-1"></h3>
+            (<span data-ng-bind="('pci_workflow_create_cron_timezone' | translate)"></span>)
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html]:
```suggestion
            (<span data-translate="pci_workflow_create_cron_timezone"></span>)
```

File: packages/manager/modules/pci/src/projects/project/workflow/workflow.controller.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      });
+  }
+
+  selectWorkflow(workflow) {
+    this.selectedWorkflowId = workflow.id;
+    this.goToExecutionsPage(workflow);
+  }
+
+  goToInstancePage(instanceId) {
+    this.$state.go('pci.projects.project.instances.instance', {
```
Comment (marie-j) [inline packages/manager/modules/pci/src/projects/project/workflow/workflow.controller.js]: We should try to pass those in the `resolve` configuration of the routing and as binding of the `workflow.component`

File: packages/manager/modules/pci/src/projects/project/sidebar/sidebar.constant.js [hunk 12]
```diff
@@ -88,12 +88,19 @@ export const MENU = [
         title: 'Managed Kubernetes Service',
       },
       {
+        beta: true,
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/sidebar/sidebar.constant.js]: Is this related to auto-backup ?

File: packages/manager/modules/pci/src/projects/project/workflow/add/general/index.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+const moduleName = 'ovhManagerPciWorkflowCreateGeneralModule';
+
+angular
+  .module(moduleName, [
+    'oui',
+    'ngTranslateAsyncLoader',
+    'pascalprecht.translate',
+  ])
+  .component('workflowGeneralInfo', component)
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/general/index.js]:
```suggestion
  .component('pciProjectsProjectWorkflowGeneral', component)
```

File: packages/manager/modules/pci/src/projects/project/workflow/add/resources/index.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+const moduleName = 'ovhManagerPciWorkflowCreateResources';
+
+angular
+  .module(moduleName, [
+    'oui',
+    'ngTranslateAsyncLoader',
+    'pascalprecht.translate',
+  ])
+  .component('workflowResources', component)
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/resources/index.js]:
```suggestion
  .component('pciProjectsProjectWorkflowAddResources', component)
```

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html (hunk 9 above)
Comment (jleveugle) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html]: Following @marie-j 's comment, you should remove the row too :)

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: you mean, code to validate UNIX Cron pattern is already exists in web-universe-components?

File: packages/manager/modules/pci/src/projects/project/sidebar/sidebar.constant.js (hunk 12 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/sidebar/sidebar.constant.js]: no, its for private registry, its added because of rebase

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.controller.js (hunk 8 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.controller.js]: this is an object, most of properties values are coming from constants.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (marie-j) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: https://github.com/ovh-ux/ng-ovh-web-universe-components/tree/master/src/cron-validator
but this implies adding it as a dependency to pci module

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 6 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: I have tried it but moment returns current date, month number if wrong data/month is provided. For example it returns 7 for moment().month('abc').format('M'). Seems like there is no way to validate it.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: can web-universe-components be added as a dependency to cloud? also I cant use it as it is. I need to make some changes to the code.

**Comment by ovh-cds** [comment]
CDS Report overthebox#3524.0 ✘  
* Stage 1
  * Build CA ✔ 
  * Build EU ✔ 
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report freefax#3524.0 ✘  
* Stage 1
  * Build CA ✔ 
  * Build EU ✘  
  * Build US ✔

**Comment by ovh-cds** [comment]
CDS Report installation#3525.0 ■   
* Stage 1
  * Install ■

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 6 above)
Comment (marie-j) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: For months, when I `moment().month('jan').format('M')` I get `1` so it seems to work fine :) 
For days you can use `format('e')` e.g `moment().day('sun').format('e')` returns `0`
https://momentjs.com/docs/#/parsing/string-format/

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 3 above)
Comment (cbourgois) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: Translations seems not loaded.

```suggestion
  .service('PciProjectRegions', PciProjectRegions)
  .run(/* @ngTranslationsInject:json ./translations */)
```

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+                </div>
+                <div class="oui-field__error pr-2" data-ng-if="$ctrl.monthInvalid">
+                    <span role="alert">
+                    <span class="oui-icon oui-icon-error" aria-hidden="true"></span>
+                        <span data-translate="pci_workflow_create_cron_invalid"></span>
+                    </span>
+                </div>
+                <div class="oui-field__helper">
+                    <span data-translate="pci_workflow_create_cron_month"></span>
+                    <oui-popover>
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html]: It should be better to set `placement="bottom"` on cron popover to avoid to hide cron fields.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+                            <ui>
+                                <li data-translate="pci_workflow_create_cron_dow_helper_1"></li>
+                                <li data-translate="pci_workflow_create_cron_dow_helper_2"
+                                    data-translate-values="{ weekNameExample: 'mon, tue, wed, thu, fri, sat, or sun' }"></li>
+                            </ui>
+                        </oui-popover-content>
+                    </oui-popover>
+                </div>
+            </div>
+            <oui-popover>
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/schedule/schedule.html]: Should be placed after the title `pci_workflow_create_cron_title`.

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 3 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: I did not see any error message on this, I verified and it looked like all required translations are loaded.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 6 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: no its not because of locale, can you please check the output of this "moment().month('test').format('M')". For me this returns current month because the month provided is wrong. If user enters a wrong month in UI, it will validate it as current month and generates wrong Cron.

**Comment by ovh-cds** [comment]
CDS Report telecom-dashboard#3648.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report freefax#3648.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report sms#3648.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  .service('PciProjectsProjectInstanceService', PciProjectsProjectInstanceService)
+  .run(/* @ngTranslationsInject:json ./translations ../../instances/translations */);
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: IMO, we should keep in mind that one module comes with it's translations and injects them.

File: packages/manager/modules/pci/src/projects/project/workflow/add/type/index.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+const moduleName = 'ovhManagerPciWorkflowTypeModule';
+
+angular
+  .module(moduleName, [
+    'oui',
+    'ngTranslateAsyncLoader',
+    'pascalprecht.translate',
+  ])
+  .component('pciProjectsProjectWorkflowType', component)
+  .run(/* @ngTranslationsInject:json ./translations */);
```
Comment (cbourgois) [inline packages/manager/modules/pci/src/projects/project/workflow/add/type/index.js]: I don't see translations file in `workflow/add/type`.

**Comment by ovh-cds** [comment]
CDS Report public-cloud#3731.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report telecom-dashboard#3743.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report sign-up#3746.0 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report sign-up#3746.1 ✘  
* Stage 1
  * Build CA ✘  
  * Build EU ✘  
  * Build US ✘

**Comment by ovh-cds** [comment]
CDS Report installation#3803.0 ✘  
* Stage 1
  * Install ✘

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 17 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: how do we reuse translations

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 3 above)
Comment (antleblanc) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: @radireddy Could you please addressed @marie-j's feedback?

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (antleblanc) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: We will benefit of not maintaining the same logic in multiple place.

I agreed with @marie-j's comment.

@radireddy Is that required a lot of change to the actual `WucCronValidator` service to fit your needs?

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 17 above)
Comment (antleblanc) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: Modules have to be well scoped. In that case it will facilitate any move from module A to another place for instance.

File: packages/manager/modules/pci/src/projects/project/workflow/delete/translations/Messages_fr_FR.json [hunk 19]
```diff
@@ -0,0 +1,9 @@
+{
+  "pci_workflow_delete_confirmation": "Êtes-vous sûr de vouloir supprimer le planning de workflow'{{nom_du_workflow}}' ?",
```
Comment (JeremyDec) [inline packages/manager/modules/pci/src/projects/project/workflow/delete/translations/Messages_fr_FR.json]:
```suggestion
  "pci_workflow_delete_confirmation": "Êtes-vous sûr de vouloir supprimer le planning de workflow '{{nom_du_workflow}}' ?",
```

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: Yes, lot of changes and testing is required. Can I address this as part of a separate issue?

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 3 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: I removed it, its not required anymore after latest changes.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+      const seconds = splits.shift();
+      if (seconds) {
+        checks.push(this.validateSeconds(seconds));
+      }
+    }
+    const minutes = splits[0];
+    const hours = splits[1];
+    const days = splits[2];
+    const months = splits[3];
+    const weekdays = splits[4];
```
[the remark this answers is not in the document]
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: each variable value is different, its not possible to declare and define all of them in single line.

File: packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js (hunk 17 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/add.module.js]: removed dependency on other module translations

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 6 above)
Comment (radireddy) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: The input field can and must accept numbers, special characters, or characters. For example the value can be * or 2 or 2-5 or 2/5 or jan. Since its very complex validation to use regex, I do the validation in code and left the input a free text which can take any value.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 7 above)
Comment (antleblanc) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: @radireddy Yes for sure.

File: packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (antleblanc) [reply packages/manager/modules/pci/src/projects/project/workflow/add/schedule/cron-validator.js]: @radireddy this is just a shorthand. It still should work as expected. This is not a mandatory but it's more readable;)
